platforms/jvm/testing-jvm-infrastructure/src/main/java/org/gradle/api/internal/tasks/testing/junit/JUnitTestEventAdapter.java

```diff
@@ -62,30 +73,174 @@ public class JUnitTestEventAdapter extends RunListener {
         new AssertErrorMapper()
     );
 
+    private static final class TestNode {
+        // Avoid generating the id too early, as it leads to a confusing ordering of ids
+        // Ordering generally shouldn't be relied on, but the tests read cleaner if ids are generated in a depth-first order
+        private final IdGenerator<?> idGenerator;
+        @Nullable
+        private volatile Object resolvedId;
+        final Description description;
+
+        TestNode(IdGenerator<?> idGenerator, Description description) {
+            this.idGenerator = idGenerator;
+            this.description = description;
+        }
+
+        public Object resolveId() {
+            Object localId = resolvedId;
+            if (localId == null) {
+                synchronized (this) {
+                    localId = resolvedId;
+                    if (localId == null) {
+                        localId = idGenerator.generateId();
+                        resolvedId = localId;
+                    }
+                }
+            }
+            return localId;
+        }
+    }
+
+    private static final class PostRunStartData {
+        final Map<Description, TestNode> parentDescToNode;
+        final Map<Description, TestNode> childDescToParent;
```

Comment:
💅 Can we clarify to _which_ nodes in the names here?
```suggestion
        final Map<Description, TestNode> parentDescToNode;
        final Map<Description, TestNode> childDescToParentNode;
```